### PR TITLE
[WIP]: Support for more types of DNS records

### LINKED ISSUE #8962: netty-resolver-dns "Got bad packet: bad label type"
I have built a simple DNS proxy server using the Netty DNS resolver library. [See the code here.](https://github.com/nzhenry/dns-proxy)

It works for most queries but not all. For example, when I try to query the A records for "www.apple.com" using nslookup or dig, it fails intermittently with this error:
```;; Got bad packet: bad label type
305 bytes
ca 20 80 00 00 01 00 04 00 00 00 01 03 77 77 77          .............www
05 61 70 70 6c 65 03 63 6f 6d 00 00 01 00 01 03          .apple.com......
77 77 77 05 61 70 70 6c 65 03 63 6f 6d 00 00 05          www.apple.com...
00 01 00 00 00 37 00 1b 03 77 77 77 05 61 70 70          .....7...www.app
6c 65 03 63 6f 6d 07 65 64 67 65 6b 65 79 03 6e          le.com.edgekey.n
65 74 00 03 77 77 77 05 61 70 70 6c 65 03 63 6f          et..www.apple.co
6d 07 65 64 67 65 6b 65 79 03 6e 65 74 00 00 05          m.edgekey.net...
00 01 00 00 30 2b 00 2f 03 77 77 77 05 61 70 70          ....0+./.www.app
6c 65 03 63 6f 6d 07 65 64 67 65 6b 65 79 03 6e          le.com.edgekey.n
65 74 0b 67 6c 6f 62 61 6c 72 65 64 69 72 06 61          et.globalredir.a
6b 61 64 6e 73 c0 41 03 77 77 77 05 61 70 70 6c          kadns.A.www.appl
65 03 63 6f 6d 07 65 64 67 65 6b 65 79 03 6e 65          e.com.edgekey.ne
74 0b 67 6c 6f 62 61 6c 72 65 64 69 72 06 61 6b          t.globalredir.ak
61 64 6e 73 03 6e 65 74 00 00 05 00 01 00 00 0c          adns.net........
b7 00 19 05 65 36 38 35 38 05 64 73 63 65 39 0a          ....e6858.dsce9.
61 6b 61 6d 61 69 65 64 67 65 c0 41 05 65 36 38          akamaiedge.A.e68
35 38 05 64 73 63 65 39 0a 61 6b 61 6d 61 69 65          58.dsce9.akamaie
64 67 65 03 6e 65 74 00 00 01 00 01 00 00 00 11          dge.net.........
00 04 68 5f a0 7e 00 00 29 10 00 00 00 00 00 00          ..h_....).......
00                                                       .
```
When it fails the debug log looks like this:
```DefaultDnsQuestion(www.apple.com. IN A)
DefaultDnsRawRecord(www.apple.com. 885 IN CNAME 27B)
DefaultDnsRawRecord(www.apple.com.edgekey.net. 10239 IN CNAME 47B)
DefaultDnsRawRecord(www.apple.com.edgekey.net.globalredir.akadns.net. 441 IN CNAME 25B)
DefaultDnsRawRecord(e6858.dsce9.akamaiedge.net. 9 IN A 4B)] RECEIVED: [[[id: 0xc9dafb10], 36804, /192.168.1.1:53, DatagramDnsResponse(from: /192.168.1.1:53, to: /0:0:0:0:0:0:0:0:51678, 36804, QUERY(0), NoError(0), RD RA)
DefaultDnsQuestion(www.apple.com. IN A)
DefaultDnsRawRecord(www.apple.com. 885 IN CNAME 27B)
DefaultDnsRawRecord(www.apple.com.edgekey.net. 10239 IN CNAME 47B)
DefaultDnsRawRecord(www.apple.com.edgekey.net.globalredir.akadns.net. 441 IN CNAME 25B)
DefaultDnsRawRecord(e6858.dsce9.akamaiedge.net. 9 IN A 4B)]: {}], {}
```
When it succeeds the log looks like this:
```DefaultDnsQuestion(www.apple.com. IN A)
DefaultDnsRawRecord(www.apple.com. 881 IN CNAME 27B)
DefaultDnsRawRecord(www.apple.com.edgekey.net. 10235 IN CNAME 50B)
DefaultDnsRawRecord(www.apple.com.edgekey.net.globalredir.akadns.net. 437 IN CNAME 28B)
DefaultDnsRawRecord(e6858.dsce9.akamaiedge.net. 5 IN A 4B)
DefaultDnsRawRecord(OPT flags:0 udp:4096 0B)] RECEIVED: [[[id: 0xc9dafb10], 37086, /192.168.1.1:53, DatagramDnsResponse(from: /192.168.1.1:53, to: /0:0:0:0:0:0:0:0:51678, 37086, QUERY(0), NoError(0), RD RA)
DefaultDnsQuestion(www.apple.com. IN A)
DefaultDnsRawRecord(www.apple.com. 881 IN CNAME 27B)
DefaultDnsRawRecord(www.apple.com.edgekey.net. 10235 IN CNAME 50B)
DefaultDnsRawRecord(www.apple.com.edgekey.net.globalredir.akadns.net. 437 IN CNAME 28B)
DefaultDnsRawRecord(e6858.dsce9.akamaiedge.net. 5 IN A 4B)
DefaultDnsRawRecord(OPT flags:0 udp:4096 0B)]: {}], {}
```
Notice the additional OPT record when it succeeds. This appears to be the key. I just don't know how I can fix the problem when it fails.

I've discovered that if I remove the CNAME records from the response, then the query resolves with no errors. I would like to return a proper copy of the server response including the CNAME records. Any help would be greatly appreciated!

### PR DESCRIPTION
**Pull request by qeesung**
Motivation:

Current DNS encoding and decoding only supports a few record types (`PTR`, `OPT`), but DNS records have many kinds, such as type `A`, `NS`, `CNAME` and so on.

Modification:

Add different types of DNS Records and codec

Result:

Fixes #8962

### CODE REVIEW

**Comment by qeesung** [comment]
Because there are dozens of different types of DNS Records, it will take some time to support them all. Currently, only `A`, `AAAA`, `NS`, `CNAME`, `PTR` types are supported.

Currently only one basic scaffold has been written, and if there are no problems, I will support more types of DNS Records.

File: codec-dns/src/main/java/io/netty/handler/codec/dns/DnsRecord.java [hunk 1]
```diff
@@ -82,4 +82,9 @@
      * Returns the time to live after reading for this resource record.
      */
     long timeToLive();
+
+    /**
+     * Decodes the record data.
+     */
+    void decodeRdata();
```
Comment (normanmaurer) [inline codec-dns/src/main/java/io/netty/handler/codec/dns/DnsRecord.java]: a Record should be just a "Pojo". Decoding does belong to a decoder and not the record

Comment (qeesung) [reply codec-dns/src/main/java/io/netty/handler/codec/dns/DnsRecord.java]: The decode operation has been moved to the decoder :-). @normanmaurer would you please review it again ?

**Comment by qeesung** [comment]
I'll shut down this PR temporarily. When all the functions are tested, I will create a new PR. @normanmaurer  thanks :-)

**Pull request closed without merging**